---
title: "End-to-End Differential Expresssion Analysis"
author: "Shane Crinion"
date: "30 May 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Library loading

Load the required libraries for the analysis:

```{r libraries}

#General Bioconductor packages
    library(Biobase)
    library(oligoClasses)
     
#Annotation and data import packages
    library(GEOquery)
    library(pd.hugene.1.0.st.v1)
    library(hugene10sttranscriptcluster.db)
     
#Quality control and pre-processing packages
    library(oligo)
    library(arrayQualityMetrics)
     
#Analysis and statistics packages
    library(limma)
    library(topGO)
    library(ReactomePA)
    library(clusterProfiler)
     
#Plotting and color options packages
    library(gplots)
    library(ggplot2)
    library(geneplotter)
    library(RColorBrewer)
    library(pheatmap)
     
#Formatting/documentation packages
   #library(rmarkdown)
   #library(BiocStyle)
    library(dplyr)
    library(tidyr)

#Helpers:
    library(stringr)
    library(matrixStats)
    library(genefilter)
    library(openxlsx)
   #library(devtools)


```


## 2. Import the data
GEO series (GSE) are related experiments that can be returned quickly as a list of ExpressionSet files using the getGEO function. The ExpressionSet data consists of assayData, metaData (such as phenoData and featureData) and experimentalData. This data can then be manipulated for use with bioConductor packages and differential expression analysis. 

```{r}

gse25462 <- getGEO("GSE25462", GSEMatrix = TRUE)
show(gse25462)

```

The data is then inspected to ensure that it the skeletal muscle samples as expected. The getGEO() query returns a list of ExpressionSet objects; this means that returns for each ExpressionSet are called accordingly, hence the use of '[[1]]'. To confirm their concordance, the GSM ID numbers are then crossvalidated on GEO under 'Samples' at the link https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse25462. 

```{r}
show(pData(phenoData(gse25462[[1]])))
```

Columns of interest include disease status (title), gender (characteristics_ch1.1), age (characteristics_ch1.2), BMI (characteristics_ch1.4) and hemoglobin a1c levels (characteristics_ch1.7).

## 3. Quality control 
This is useful for checking for outliers and identifying clusters depending on their disease status. Each row represents a single microarray probe, while the columns represent a one sample from the series. 

```{r}
Biobase::exprs(gse25462[[1]])[1:5, 1:50]
```

The log2 is taken of the above as quality control and as is standard for expression data. The principal component analysis (PCA) is plotted also to indicate the sample variation. 

((--- unnecessary) As shown above by the phenotype data, the samples indicate the disease state in the 'title' column but are incompatible for grouping due to their numbered format (eg. Muscle-FamilyHistoryNegative-1, Muscle-FamilyHistoryNegative-2). The samples labels are edited before PCA for compatibility: ((--- unnecessary)))

```{r}
#extract the nested list
#gse25462 <- gse25462[[1]]

#select the corresponding samples for each disease state
#for (i in gse25462$title) {
#  i[c(1:13,47,49)] <- "FH-" # family history neg 
#  i[c(14:36,48,50)] <- "FH+" # family history pos
#  i[37:46] <- "T2D" # diabetes 
#  #convert to factor to replace the numbered with fixed labels 
#  i <- as.factor(i)
#}

#gse25462$title <- i
#gse25462$title
```

PCA is first performed using fasting glucose, however the fasting glucose levels, the results were quite messy.

```{r}

# make the empty column 
gse25462[[1]]$insulin_category <- 0

# assign each column to its appropriate bin 
gse25462[[1]]$insulin_category[(as.numeric(gse25462[[1]]$characteristics_ch1.9))>=8] <- "diabetic"

gse25462[[1]]$insulin_category[((as.numeric(gse25462[[1]]$characteristics_ch1.9))< 8 
                               & (as.numeric(gse25462[[1]]$characteristics_ch1.9))> 3)] <- "optimal"

gse25462[[1]]$insulin_category[(as.numeric(gse25462[[1]]$characteristics_ch1.9))< 3] <- "low"


#log 2
exp_raw <- log2(Biobase::exprs(gse25462[[1]]))
#pca
PCA_raw <- prcomp(t(exp_raw), scale. = FALSE)


percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    Disease = pData(gse25462[[1]])$characteristics_ch1.3,# disease state
                    Phenotype = pData(gse25462[[1]])$insulin_category, #fasting glucose levels
                    Individual = pData(gse25462[[1]])$title)

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(
        aes(shape = Disease, 
            colour = Phenotype)) +
  
  ggtitle("PCA plot of the log-transformed raw expression data") +
  
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  
  theme(plot.title = element_text(hjust = 0.5))+
  
  coord_fixed(ratio = sd_ratio) 
#  scale_shape_manual(values = c(4,15)) + 
# scale_color_manual(values = c("darkorange2", "dodgerblue4", "gold", "deepskyblue1"))

```


Given that Synthea data contains many records for H1Ac levels, it may be useful to identify the association between Ha1c levels and diabetes status. The windows for diabetic levels Hemoglobin A1c (HbA1c) are < 6% is considered healthy, between 6% and 6.4% is pre-diabetic and 6.5% or above is diabetic. 
```{r}
# make the empty column 
gse25462[[1]]$diabetes_status <- 0

# assign each column to its appropriate bin 
gse25462[[1]]$diabetes_status[gse25462[[1]]$`hemoglobin a1c:ch1`>=6.5] <- "diabetic levels"

gse25462[[1]]$diabetes_status[(gse25462[[1]]$`hemoglobin a1c:ch1`< 6.5 
                               & gse25462[[1]]$`hemoglobin a1c:ch1`> 6)] <- "pre-diabetic levels"

gse25462[[1]]$diabetes_status[gse25462[[1]]$`hemoglobin a1c:ch1`< 6] <- "normal levels"


dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    Disease = pData(gse25462[[1]])$characteristics_ch1.3,# disease state
                    Phenotype = pData(gse25462[[1]])$diabetes_status, #fasting glucose levels
                    Individual = pData(gse25462[[1]])$title)

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(
        aes(shape = Disease, 
            colour = Phenotype)) +
  
  ggtitle("PCA plot of the log-transformed raw expression data") +
  
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  
  theme(plot.title = element_text(hjust = 0.5))+
  
  coord_fixed(ratio = sd_ratio) 

```

The PCA plot shows the log intensity scale calculated from the raw data. It indicates that the first principal component differentiates between the samples, indicating that the disease type is a driver of gene expression differences. The results however do show a lack of concordance among the H1ac measurements with diabetes diagnosis. 


The probe intensities can also be shown by their individual microarray using a boxplot:

```{r}
oligo::boxplot(gse25462[[1]], target = "core", 
               main = "Boxplot of log2-intensitites for the raw data")
```


The graph indicates that there is a high level of variation, especially on the further end. A more elaborate report of the array quality can be generated usign arrayQualityMetrics


```{r}

arrayQualityMetrics(expressionset = gse25462[[1]],
    outdir = tempdir(),
    force = TRUE, do.logtransform = TRUE,
    intgroup = c("characteristics_ch1.3", "characteristics_ch1.8"))
```

## Background adjustments
Summarisation

```{r}

head(ls("package:hugene10sttranscriptcluster.db"))
```


Normalisation: 
The normalisation of an ExpressionSet can be perfomed using the package affyPLM below. The transfn function is useful when dealing with the log values. 

```{r}
library(affyPLM)
gse_norm <- normalize(gse25462[[1]], transfn=c("log"))

```

Plotting the Relative Log Expression

```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(gse_norm)))

RLE_data <- sweep(Biobase::exprs(gse_norm), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  ylim(c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))
```

The above indicates that there are no batch effects as seen by the lack of y-directed deviation and therefore no outliers need to be removed. 

